File: native/ios/NEMADataBridge.swift
import Foundation

/// Native ingress contract for NEMA Drive.
/// The host app supplies real SDK adapters (HERE Navigate, Vision/Core ML, vehicle data, etc.).
public protocol NEMADataAdapter {
    var name: String { get }
    var capabilities: [String: Bool] { get }
}

public protocol NEMATrafficAdapter: NEMADataAdapter {
    func fetchTraffic(latitude: Double, longitude: Double, completion: @escaping (Result<[String: Any], Error>) -> Void)
}

public protocol NEMAVisionAdapter: NEMADataAdapter {
    func detect(frame: Any, completion: @escaping (Result<[[String: Any]], Error>) -> Void)
}

public protocol NEMAOfflineRoutingAdapter: NEMADataAdapter {
    func route(from: (lat: Double, lon: Double), to: (lat: Double, lon: Double), completion: @escaping (Result<[String: Any], Error>) -> Void)
}

public protocol NEMA3DMapAdapter: NEMADataAdapter {
    func loadScene(latitude: Double, longitude: Double, completion: @escaping (Result<[String: Any], Error>) -> Void)
}

public final class NEMADataBridge {
    public static let shared = NEMADataBridge()
    private var adapters: [String: NEMADataAdapter] = [:]

    private init() {}

    public func register(_ adapter: NEMADataAdapter) {
        adapters[adapter.name] = adapter
    }

    public func status() -> [String: Any] {
        var result: [String: Any] = [:]
        adapters.forEach { result[$0.key] = $0.value.capabilities }
        return ["registered": result, "count": adapters.count]
    }

    /// Send a normalized event to the web/runtime layer.
    /// The WKWebView bridge should forward this as a `window.postMessage` or
    /// evaluate JavaScript against `NEMARealData`.
    public func event(_ name: String, payload: [String: Any]) -> [String: Any] {
        return ["name": name, "payload": payload, "timestamp": Date().timeIntervalSince1970 * 1000]
    }
}

File: native/android/NemaDataBridge.kt
package com.nemadrive.nativebridge

/** Native ingress contract for NEMA Drive.
 * Host app adapters provide real SDK implementations.
 */
interface NemaDataAdapter {
    val name: String
    val capabilities: Map<String, Boolean>
}

interface NemaTrafficAdapter : NemaDataAdapter {
    suspend fun fetchTraffic(latitude: Double, longitude: Double): Map<String, Any?>
}

interface NemaVisionAdapter : NemaDataAdapter {
    suspend fun detect(frame: Any): List<Map<String, Any?>>
}

interface NemaOfflineRoutingAdapter : NemaDataAdapter {
    suspend fun route(fromLat: Double, fromLon: Double, toLat: Double, toLon: Double): Map<String, Any?>
}

interface Nema3DMapAdapter : NemaDataAdapter {
    suspend fun loadScene(latitude: Double, longitude: Double): Map<String, Any?>
}

object NemaDataBridge {
    private val adapters = linkedMapOf<String, NemaDataAdapter>()

    fun register(adapter: NemaDataAdapter) {
        adapters[adapter.name] = adapter
    }

    fun status(): Map<String, Any> = mapOf(
        "registered" to adapters.mapValues { it.value.capabilities },
        "count" to adapters.size
    )

    fun event(name: String, payload: Map<String, Any?>): Map<String, Any?> = mapOf(
        "name" to name,
        "payload" to payload,
        "timestamp" to System.currentTimeMillis()
    )
}
